### PR TITLE
Make LLVM Profiling robust for multithreaded programs

### PR DESCRIPTION
**Pull request by bachdavi**
This is a draft PR to make LLVM optimization logging correct for multithreaded programs.

It guards against the case where you enable or disable LLVM optimization logging (`@snoopl`) in one thread, while another thread is midway through an ongoing optimization, or stopping it before it's finished. Without this PR, this can result in a malformed .yaml file, since the logging might start with the "after" clause, or it might only contain the "begin" clause.

After this PR, such cases will simply not be logged, and only LLVM optimizations that occurred completely while logging was enabled would be logged.

### CODE REVIEW

**Comment by pchintalapudi** [comment]
It's not super clear to me what the purpose of this PR is. Is the intent to allow access to these timings within Julia code? If so, why is the timing struct embedding a vector<> and not a julia array? Also, the previous implementation of timing reporting should have been thread-safe, albeit not manipulable from Julia, since we locked before ever operating on the stream.

**Comment by bachdavi** [comment]
@pchintalapudi This is very much a work in progress :) 
I pushed a PR such that @NHDaly and @vilterp can work on this together. 
We'll ping you once this is in any state to review it!

**Comment by bachdavi** [comment]
Exactly, we want to access the timings from within Julia while potentially having another thread do code generation. Interesting, we thought that we might run into a case where we unset the streams, and codgen is trying to write to it, finding a null pointer. Where does the locking happen?

File: src/codegen.cpp [hunk 1]
```diff
@@ -8133,17 +8135,7 @@ jl_llvm_functions_t jl_emit_code(
         "functions compiled with custom codegen params must not be cached");
     JL_TRY {
         decls = emit_function(m, li, src, jlrettype, params);
-        auto stream = *jl_ExecutionEngine->get_dump_emitted_mi_name_stream();
```
Comment (pchintalapudi) [inline src/codegen.cpp:8136]: This "dereference" is actually overloaded to take the stream's lock (defined in `jl_locked_stream` in jitlayers.h) and then release when the scope block ends.

Comment (NHDaly) [reply src/codegen.cpp:8136]: fascinating. that was super non-obvious from the call site, but that's really good to know!

Comment (NHDaly) [reply src/codegen.cpp:8136]: Thanks Prem!

Comment (pchintalapudi) [reply src/codegen.cpp:8136]: Note that this dereference locking is also used in the debuginfo code as well when interacting with the global registry.

**Comment by bachdavi** [comment]
@pchintalapudi Does the actual optimization in `jitlayers.cpp`: `(***PMs).run(M);` acquire any locks? If not, would it be feasible to hold the stream lock over the whole optimization time to ensure we don't write just the first part of the YAML block or just the second one? I've adapted this PR to do that. The alternative would be to write the first part into a string and write the whole YAML block atomically after optimization.

**Comment by NHDaly** [comment]
🤔 @bachdavi: now that i'm looking at the PR change, we maybe don't want to hold the lock over the whole LLVM optimization, since that would force serialization of two different threads doing LLVM optimization if we were profiling..

So i think that the string-buffer approach is probably the best one. 👍 Sorry i didn't think of this when we discussed it earlier! Thanks for the quick change today :)

**Comment by pchintalapudi** [comment]
We actually specifically want to avoid holding a lock during optimization, so that in the future multiple modules can be optimizing at the same time. Rather than write to a string first, I would say holding the data in variables and dumping the entire yaml block post-optimization using those variables would be a better solution.

**Comment by NHDaly** [comment]
💡 Prem, yeah, just keeping the results in variables seems even better. 👍 👍

**Comment by pchintalapudi** [comment]
As an aside, we currently don't measure machine code optimization and emission time at the moment (this is in jitlayers.cpp, CompilerT::operator()) although this probably constitutes a decent fraction of optimization time. That may also be something of interest if you're looking to collect optimization statistics.

**Comment by NHDaly** [comment]
I was wondering about that, @pchintalapudi: Does that get measured in `@snoopc` (codegen)?

**Comment by pchintalapudi** [comment]
It does get measured by `@snoopc` since that relies on the `dump_compiles_stream` instead of the `dump_llvm_opt_stream`, but that includes time spent in codegen, middle-end optimization, and backend optimization/mc emission. We don't currently have anything measuring the contribution of just backend optimization/mc emission.

**Comment by bachdavi** [comment]
This uses a `stringstream`, instead we could use a `map` and store `instructions` and `basicblocks` as a tuple. The downside of the `stringstream` is the lack of proper formatting, afaik `std::format` is only available from c++20.

File: src/jitlayers.cpp [hunk 2]
```diff
@@ -1123,6 +1127,7 @@ namespace {
                 {
                     auto stream = *jl_ExecutionEngine->get_dump_llvm_opt_stream();
                     if (stream) {
+                        jl_printf(stream, "%s", before_stats_ss.str().c_str());
```
Comment (NHDaly) [inline src/jitlayers.cpp:1130]: You'll need to check here that the before string isn't empty, to catch the case where this is enabled halfway through.
Maybe you could just add a separate boolean that starts out false and gets set to true inside the first block?

**Review by NHDaly — APPROVED** [review]
LGTM, thanks, @bachdavi! 
I've updated the description as well, to explain this situation better.

@pchintalapudi: Can you do a final review pass as well?

**Comment by NHDaly** [comment]
Friendly ping for @pchintalapudi. Do you want to review? We've been using this in production at RAI and so far it seems to be working correctly. Can we merge this?

CC: @IanButterworth as well, since this is profiling related.

**Review by pchintalapudi — APPROVED** [review]
Seems fine to me

**Comment by vchuravy** [comment]
Needs a rebase

**Comment by NHDaly** [comment]
Also, are the test failures here real? Or unrelated?

**Comment by vchuravy** [comment]
`analyzegc` failure seems relevant.

**Comment by bachdavi** [comment]
> `analyzegc` failure seems relevant.

Mh, it seems so indeed!

**Comment by NHDaly** [comment]
Thanks Valentin and David. The remaining failures look unrelated:
```
Error During Test at C:\buildkite-agent\builds\win2k22-aws-vms-2\julialang\julia-master\julia-4d55f71f4d\share\julia\stdlib\v1.10\Downloads\test\runtests.jl:12
  Got exception outside of a @test
  RequestError: Could not resolve host: httpbingo.julialang.org while requesting https://httpbingo.julialang.org/base64/SnVsaWEgaXMgZ3JlYXQh
  Stacktrace:
    [1] (::Downloads.var"#9#18"{IOStream, Base.DevNull, Nothing, Vector{Pair{String, String}}, Float64, Nothing, Bool, Nothing, Bool, String, Bool, Bool})(easy::Downloads.Curl.Easy)
```
and
```
intrinsics                                       (5) |        started at 2023-01-09T13:12:37.840
      From worker 4:
      From worker 4:	[180] signal (11.1): Segmentation fault
      From worker 4:	in expression starting at /cache/build/default-power9bot2-0/julialang/julia-master/julia-4d55f71f4d/share/julia/test/atomics.jl:281
```

**Pull request merged**